res/controllers/Denon-Prime-4-scripts.js

```diff
@@ -0,0 +1,412 @@
+        key: "slip_enabled",
+        type: components.Button.prototype.types.toggle,
+    });
+    // Loop In Button
+    this.loopInButton = new components.Button({
+        midi: [0x90 + midiChannel, 0x25],
+        key: "loop_in",
+    });
+    // Loop Out Button
+    this.loopOutButton = new components.Button({
+        midi: [0x90 + midiChannel, 0x26],
+        key: "loop_out",
+    });
+
+    this.deckLoad = new components.Button({
+        midi: [0x9F, midiChannel - 3],
+        key: "LoadSelectedTrack",
+
+        // TODO: Make load buttons reflect colour of currently active deck.
+        //       This is harder than expected due to the fact that I rely on the MIDI channel
+        //       of each physical deck component to determine which deck it's manipulating in
+        //       Mixxx. The deck load buttons, however, are both on MIDI channel 16, so I need
+        //       a better way to determine this.
+        off: Prime4.rgbCode.greenDark,
+        on: Prime4.rgbCode.green,
+    });
+
+    // Change from Deck 3/4 to Deck 1/2
+    this.deckToggleButtonA = function(value) {
+        if (value > 0) {
+            // Each physical deck has its own MIDI channel, which channelCheck converts to the appropriate deck
+            const channelCheck = midiChannel - 1;
+            if (this.currentDeck === "[Channel" + channelCheck + "]") {
+                this.toggle();
+
+                // Update Jog Wheel and Toggle Button LEDs to reflect active deck
+                midi.sendShortMsg(0x90 + midiChannel, 0x21, colDeck[midiChannel - 4]); // Jog Wheel LED
+                midi.sendShortMsg(0x9F, 0x18 + midiChannel, colDeck[midiChannel - 4]); // Active deck - bright LED
+                midi.sendShortMsg(0x9F, 0x1A + midiChannel, colDeckDark[midiChannel - 2]); // Inactive deck - dark LED
+            }
+        }
+    };
+    // Change from Deck 1/2 to Deck 3/4
+    this.deckToggleButtonB = function(value) {
+        if (value > 0) {
+            const channelCheck = midiChannel - 3;
+            if (this.currentDeck === "[Channel" + channelCheck + "]") {
+                this.toggle();
+
+                // Update Jog Wheel and Toggle Button LEDs to reflect active deck
+                midi.sendShortMsg(0x90 + midiChannel, 0x21, colDeck[midiChannel - 2]); // Jog Wheel LED
+                midi.sendShortMsg(0x9F, 0x18 + midiChannel, colDeckDark[midiChannel - 4]); // Inactive deck - dark LED
+                midi.sendShortMsg(0x9F, 0x1A + midiChannel, colDeck[midiChannel - 2]); // Active deck - bright LED
+            }
+        }
+    };
+
+    this.reconnectComponents(function(c) {
+        if (c.group === undefined) {
+            c.group = this.currentDeck;
+        }
+    });
+};
+
+Prime4.Deck.prototype = new components.Deck();
+
+// View Button
+Prime4.maxView = new components.Button({
+    midi: [0x9F, 0x07],
+    group: "[Master]",
+    key: "maximize_library",
+    type: components.Button.prototype.types.toggle,
+});
```

Comment:
This belongs into `Prime4.init` IMO. As a rule of thumb: Everything that uses midi or ControlObject (CO) interaction should be put into the `.init`, everything that doesn't, belongs outside. So when you "instantiate" the maximize button, it asks mixxx for the value of `maximize_library` (so it reads a CO) and uses that to initialize the LED feedback of that button (so it uses midi output). For that reason, you should delay the creation/instantiation to the `.init`. As a counterexample, when you overwrote the `components.Button.isPress` function, you only declared a function and no midi or CO access was taking place, as such, it belongs outside the `.init` function.

Reply:
Right, that makes sense. Cheers